orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (graph.IsInputsIncludingInitializers(&node_arg)) {
+          connected_inputs.insert(&node_arg);
+        } else {
+          const Node* producer_node = graph.GetProducerNode(node_arg.Name());
+          if (producer_node == nullptr) {
+            // got nullptr as producer node. This could be because the input is a constant op which will be optimized
+            // away. Print out this information and continue.
+            LOGS_DEFAULT(WARNING) << "Cannot find producer node for node_arg: " << node_arg.Name() << ". Skipping this node.";
+          } else {
+            connected_nodes.push_back(producer_node);
+          }
+        }
+        return Status::OK();
+      }));
+
+  ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+      node->OutputDefs(),
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (!graph.IsOutput(&node_arg)) {
+          std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(node_arg.Name());
+          connected_nodes.insert(std::end(connected_nodes), consumer_nodes.begin(), consumer_nodes.end());
+
+        } else {
+          connected_outputs.insert(&node_arg);
+        }
+        return Status::OK();
+      }));
+  return Status::OK();
+}
+
+// PipelineStageNodeGroup groups nodes that share the same input initializer and belong to the same stage.
+// It is used to distinguish other nodes that share the same input initializer but belong to
+// other pipeline partitions after split.
+struct PipelineStageNodeGroup {
+  const size_t stage_id;
+
+  // Vector of nodes that have the same initializer input and belong to the same stage. Noted that
+  // the consumer nodes of a particular initializer can be more than one, so we need a vector to store those
+  // nodes.
+  std::vector<Node*> nodes;
+  PipelineStageNodeGroup(const size_t stage, std::vector<Node*>& node) : stage_id(stage), nodes(std::move(node)){};
+};
+
+common::Status AddPassthroughInitializer(Graph& graph,
+                                         const NodeArg* initializer,
+                                         const std::vector<PipelineStageNodeGroup>& node_groups,
+                                         const std::vector<Node*>& send_nodes,
+                                         const std::vector<Node*>& recv_nodes) {
+  ORT_ENFORCE(node_groups.size() >= 2, "Initializer ", initializer->Name(), " is not shared across stages.");
+
+  const size_t from_stage = node_groups.front().stage_id;
+  const size_t to_stage = node_groups.back().stage_id;
+
+  ORT_ENFORCE(from_stage < to_stage, "Pass through from_stage (", from_stage,
+              ") is not less than the to_stage (", to_stage, ").");
+
+  auto dtype = initializer->TypeAsProto()->tensor_type().elem_type();
+
+  // new_node_args tracks newly created node_args in the pass through stages
+  std::vector<NodeArg*> new_node_args;
+  auto current_node_arg = const_cast<NodeArg*>(initializer);
```

Comment:
Is casting away const the right approach here -- what happens if current-node_arg was made "const NodeArg *"?

Reply:
Good question. The reason for casting away const is because we need to pass it in as send node's new input (line 898 `MutableInputDefs()`). That method only accept mutable NodeArg*. Under the hood, initializer is stored as none-const NodeArg* in the graph. (actually all NodeArgs store in graph are none-const so they are ok to be modified when needed). When no modification is intended, graph.h provides another const container wrapper for that purpose. That is used earlier when we traverse the graph to get each partition nodes (and find the shared initializer).


---
In reply to: [452264705](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452264705) [](ancestors = 452264705)